Development branch

LivingTribunal18 opened the pull request.

LivingTribunal18 committed:
Update README.md

LivingTribunal18 committed:
command generator for vm

based on tokens generates commands for a virtual machine

LivingTribunal18 committed:
Update README.md

LivingTribunal18 committed:
Merge branch 'master' into development-branch

arcturus5340 commented on lang.cpp:
```diff
@@ -27,68 +35,111 @@ bool isSeparator(char list){
-   for(int i = 0; i < parsing.size(); i++){
-        for (int j = 0; j < transfer.size(); j++) {
-          cout << parsing[i][j] << "\n";
+vm_tmp.push_back("Identifier");
```
Why do You need to add tokens to vm_tmp?

arcturus5340 commented on lang.cpp:
```diff
@@ -27,68 +35,111 @@ bool isSeparator(char list){
+
+int counter = 0;
+
+for(int i = 0; i < code_tokens.size(); i++){
```
O(n^6)? It's too loong. You can check vector of tokens with vector of tokens-templates in O(n^2) + find()

arcturus5340 commented on lang.cpp:
```diff
@@ -27,68 +35,111 @@ bool isSeparator(char list){
-        word += ch;
-      }
+  string word = "";
+  while((ch = getc(code)) != EOF){
```
You need to add a variable reset in the function, since this action is repeated many times

arcturus5340 commented:
I don't see your reaction. Maybe, some problems on github?

LivingTribunal18 commented:
most likely there are possible bugs on the githab, since I responded to your review long ago

LivingTribunal18 commented:
Can you please describe, how i should properly use the find() function?

arcturus5340 commented:
`InputIterator find (InputIterator first, InputIterator last, const T& val);`
See more on http://www.cplusplus.com/reference/algorithm/find/

LivingTribunal18 commented:
Where and how do i need to implement function find() to speed up my command generator for vm?

arcturus5340 commented:
You can do this with just two nested loops. The first will iterate all the tokens, and the second all templates and, in the case of equality, bytecode will be generated. To insert literals into bytecode, you need to select them from the main-source-code. To do this, you need to know their index, and the search function will help You with this.

LivingTribunal18 commented:
How will the byte-code be generated in case of equality?

arcturus5340 commented:
You need to write bytecode commands to the file. See https://github.com/arcturus5340/MY_SVM/blob/master/make.py for details.

LivingTribunal18 committed:
second command generator for vm

Accelerated command generation cycle. 
There is no comparison with patterns, but the command generation goes right away.

arcturus5340 commented:
Nop. First, it may not work for all cases. Secondly, you don’t have a separation between parsing, tokenizer and code generator. This is bad.

arcturus5340 commented on lang.cpp:
```diff
@@ -63,8 +62,16 @@ void push(string word){
      }else if(isLiteral(word)){
        tmp_tokens.push_back("Literal");
-    }else{
+       s.push_back("PUSH " + word);
```
Some patterns can be repeated and the literal is not always a sign of the command PUSH

arcturus5340 commented on lang.cpp:
```diff
@@ -100,45 +107,13 @@ int main(){
   }
 
 
-vector<vector<string> > vm_comands;
```
It would be better to break the code into functional fragments, and not be confused to fit everything into one action. It is required to separate syntax, semantic analysis and code generation

LivingTribunal18 committed:
byte_code generator

Generation of byte code is accelerated. Byte code is written to a separate file. Generation is separated from tokenizer and parsing

arcturus5340 commented on lang.cpp:
```diff
@@ -108,12 +103,41 @@ int main(){
+vm_comands.push_back(vm_tmp);
+
+int counter = 0;
+string tok = "Literal"; string tok1 = "Identifier";
```
Do you create a new string variable to use this only once?!

arcturus5340 commented on lang.cpp:
```diff
@@ -108,12 +103,41 @@ int main(){
+         counter++;
+      }
+      if (counter >= vm_comands.size()) {
+           int it = find(code_tokens[i].begin(), code_tokens[i].end(), tok) - code_tokens[i].begin();
```
Variable names leave much to be desired..

arcturus5340 commented on lang.cpp:
```diff
@@ -108,12 +103,41 @@ int main(){
+      if (counter >= vm_comands.size()) {
+           int it = find(code_tokens[i].begin(), code_tokens[i].end(), tok) - code_tokens[i].begin();
+           int it1 = find(code_tokens[i].begin(), code_tokens[i].end(), tok1) - code_tokens[i].begin();
+           vm_code << "PUSH " << code_list[i][it] << "\n";
```
wtf? I sad, You need to give me only byte code in special *.jt file. Something like this: [one.jt.tar.gz](https://github.com/LivingTribunal18/airy/files/2813127/one.jt.tar.gz)
Output of ```hexdump -C one.jt```:
```
00000000  0a ff f0 0a ff fe 0d 0e  00                       |.........|
00000009
```

LivingTribunal18 commented:
A binary file can be a file with the extension * txt

arcturus5340 commented:
So what?

LivingTribunal18 committed:
Add files via upload

Writing the generated code to a binary file. Variable names changed. Changed counter implementation

arcturus5340 commented on lang.cpp:
```diff
@@ -113,28 +113,34 @@ int main(){
-vm_tmp.push_back("Operator");
-vm_tmp.push_back("Literal");
-vm_comands.push_back(vm_tmp);
+vm_tmp.push_back("Identifier"); vm_tmp.push_back("Operator");
```
Earlier the arrangement of commands was correct in terms of formatting

arcturus5340 commented on lang.cpp:
```diff
@@ -113,28 +113,34 @@ int main(){
 
-int counter = 0;
-string tok = "Literal"; string tok1 = "Identifier";
+int push_byte = 0x03;
```
It is more correct to use type char

arcturus5340 reviewed (changes requested):
I don't understand, why You need vector<int> counter?

LivingTribunal18 committed:
Add files via upload

Fixed writing to a binary file.

LivingTribunal18 commented:
Because we will have a lot of templates. The size of the vector is equal to the number of templates. For example, 0 index of counter is the counter of the first pattern. This is more convenient than using multiple variables in the form of a counter for each template.

LivingTribunal18 committed:
byte-code generator for vm

Fixed command layout in terms of formatting.
Int type has been replaced by char type.

arcturus5340 commented:
Why do You need counter for each template?

arcturus5340 commented on lang.cpp:
```diff
@@ -70,6 +71,36 @@ void push(string word){
 }
 
+int asci = 0;
+void ascii(string letter){
```
This code is better:
```int ascii = (int)letter[0];```

LivingTribunal18 commented:
int ascii = (int)letter[0];
I tried this option. But unfortunately he is not working.

LivingTribunal18 commented:
I need a counter for each template so that when comparing the token vector of tokens with a set of templates to find matches.
